lib/ts/Diags.cc

```diff
@@ -226,43 +214,50 @@ void
 Diags::print_va(const char *debug_tag, DiagsLevel diags_level, const SourceLocation *loc, const char *format_string,
                 va_list ap) const
 {
-  struct timeval tp;
-  const char *s;
-  char *buffer, *d, timestamp_buf[48];
-  char format_buf[1024], format_buf_w_ts[1024], *end_of_format;
-
   ink_release_assert(diags_level < DiagsLevel_Count);
 
-  ////////////////////////////////////////////////////////////////////////
-  // there are 2 format buffers that hold a printf-style format string  //
-  // format_buf contains <prefix_string>: (<debug_tag>) <format_string> //
-  // and format_buf_w_ts has the same thing with a prepended timestamp. //
-  ////////////////////////////////////////////////////////////////////////
+  const size_t BUF_CAPACITY = 1024;
+  using ts::LocalBufferWriter;
+  LocalBufferWriter<BUF_CAPACITY> format_writer;
 
-  format_buf[0]      = NUL;
-  format_buf_w_ts[0] = NUL;
+  // Save room for optional newline and terminating NUL bytes.
+  format_writer.clip(2);
 
-  /////////////////////////////////////////////////////
-  // format_buf holds 1024 characters, end_of_format //
-  // points to the current available character       //
-  /////////////////////////////////////////////////////
+  //////////////////////
+  // append timestamp //
+  //////////////////////
+  {
+    struct timeval tp = ink_gettimeofday();
+    time_t cur_clock  = (time_t)tp.tv_sec;
+    char timestamp_buf[48];
+    char *buffer = ink_ctime_r(&cur_clock, timestamp_buf);
 
-  end_of_format  = format_buf;
-  *end_of_format = NUL;
+    snprintf(&(timestamp_buf[19]), (sizeof(timestamp_buf) - 20), ".%03d", (int)(tp.tv_usec / 1000));
 
-  // add the thread id
-  end_of_format += snprintf(end_of_format, sizeof(format_buf), "{0x%" PRIx64 "} ", (uint64_t)ink_thread_self());
+    format_writer.write('[');
+    format_writer.write(buffer + 4, std::min(strlen(buffer + 4), sizeof(timestamp_buf) - 4));
```

Comment:
Why not just `strlen`? If `snprintf` went over the buffer length, you're already toast. The call to `strlen` will explode anyway. It seems safer to capture the return value of `snprintf` as you do elsewhere and work with that vs. `sizeof(timestamp_buf)`.

Reply:
Yes I agree with that. It seems better to check the result of `snprintf` here.